### DISCUSSION TITLE
Wiring up laser psu with active L fire button and 0-5v PWM Intensity

### DISCUSSION THREAD

**Discussion by jerseyguy1996** [grblHAL/core#62] [General]
Trying to wire up a new laser PSU.  I believe when I last did this I used one pin on the teensy to enable the laser and a PWM pin to control the intensity.  I believe the gcode from lightburn would just enable the laser once at the beginning of the file and then control firing the laser using S values.  Now I'm starting with a fresh grblhal install and I can't remember how I set it up.  I see on the config file the following:

`// NOTE: If VARIABLE_SPINDLE is enabled(default), this option has no effect as the PWM output and
// spindle enable are combined to one pin. If you need both this option and spindle speed PWM,
// uncomment the config option USE_SPINDLE_DIR_AS_ENABLE_PIN below.`

however, I can't find USE_SPINDLE_DIR_AS_ENABLE_PIN anywhere else in the repo.  How should I be wiring it up?  This is the PSU I'm using.  The pinouts are described.

[https://www.cloudraylaser.com/collections/power-supply/products/cloudray-35-50w-myjg-ng-co2-laser-power-supply?variant=32921182765196](url)

**Comment by terjeio** [comment]
Which pin is used for what is neither known by the core or defined by it - you have to consult the board map file you are using or look at the output from the new `$pins` command (if supported by the driver).

The USE_SPINDLE_DIR_AS_ENABLE_PIN comment is a remnant from the Grbl codebase that I have overlooked. I will remove it in the next build.

**Reply by jerseyguy1996** [reply]
@terjeio thank you for your response.  The board is Phil Barret's T41IU5XBB breakout which shows the spindle pin mappings as

![image](https://user-images.githubusercontent.com/4678592/130619339-22c27753-77c2-4f58-b50e-4f72d692f55e.png)

I guess my question relates to the behavior of the spindle outputs when VARIABLE_SPINDLE is enabled.  It says that PWM and EN are combined on one pin so does the PWM signal come out of the PWM pin above or the EN pin above?  If that's the case and I want to control the laser with 2 pins , one for EN which will idle high (or float?) and then go low when an M3 or M4 command is received, and one for intensity which will be a PWM signal dictate by a G1 S command, how would I do that?  Am i even thinking about this the right way?  I know that I used two pins in my last grblhal build, I just can't remember how I set it up.

**Reply by terjeio** [reply]
Enable and PWM signals are assigned to two different pins - 12 and 13. Enable is driven high (M3 or M4) or low (M5) unless inverted by a flag in the $16 setting. The PWM signal is switch off with a M5 unless $35 is set > 0. Setting $32 to 1 enables [laser mode](https://github.com/gnea/grbl/wiki/Grbl-v1.1-Laser-Mode) where G0 motions disables laser output.

FYI the iMXRT1061 driver supports [PPI mode](https://github.com/grblHAL/Plugins_laser) (enable in my_machine.h), I use this most of the time with my CO2 laser to get rid of overburning in corners.

**Reply by jerseyguy1996** [reply]
Thank you!  Turns out I was using the wrong firmware version and my pins were not what I thought they were.  Everything is working great now!

**Comment by customsolutions** [comment]
not sure this is the correct place, I've run into this before but cant find or remember the fix, I need to use pb9 for pwm, and when I pu that in my board file, I get timer errors. If I'm seeing and remembering correctly it has to do with driver.h Here:
 #elif SPINDLE_PWM_PORT_BASE == GPIOB_BASE
  #if SPINDLE_PWM_PIN == 0 // PB0 - TIM1_CH2N
    #define SPINDLE_PWM_TIMER_N     1
    #define SPINDLE_PWM_TIMER_CH    2
    #define SPINDLE_PWM_TIMER_INV   1
    #define SPINDLE_PWM_AF_REMAP    0b01
  #endif

Again not 100% sure the fix. If you can point me in the correct direction or help me remember.
thx

**Reply by terjeio** [reply]
An entry for  PB9 has to be added, it is bound to timer 4 channel 4. And the debounce timer has to be changed away from timer 4.

**Comment by customsolutions** [comment]
May i ask how you did that.... Am I looking in the wrong place? When I
initially did a build with spindle pwm on pb9 i recieved all kinds of
errors. Related to timer. Then I added an enrty below
elif spindle_pwm_port_base == gpiob_base for pin ==bp9 and they went away,
but havent set tim correctly.
Where did you do the change, i dont see it.(trying to learn).

**Reply by terjeio** [reply]
I did not add the code, I only suggested how. However, you may try this:

```
...
#elif SPINDLE_PWM_PORT_BASE == GPIOB_BASE
  #if SPINDLE_PWM_PIN == 0 // PB0 - TIM1_CH2N
    #define SPINDLE_PWM_TIMER_N     1
    #define SPINDLE_PWM_TIMER_CH    2
    #define SPINDLE_PWM_TIMER_INV   1
    #define SPINDLE_PWM_AF_REMAP    0b01
  #elif SPINDLE_PWM_PIN == 9 // PB9 - TIM4_CH4
    #define SPINDLE_PWM_TIMER_N     1
    #define SPINDLE_PWM_TIMER_CH    4
    #define SPINDLE_PWM_TIMER_INV   0
    #define SPINDLE_PWM_AF_REMAP    0
  #endif
#endif
```

I am not able to verify this until I am back home so I leave that to you for now. And do not forget to change the debounce timer to an unused one.

**Comment by customsolutions** [comment]
SPINDLE_PWM_PIN == 9 // PB9 - TIM4_CH4
    #define SPINDLE_PWM_TIMER_N     4
    #define SPINDLE_PWM_TIMER_CH    4
    #define SPINDLE_PWM_TIMER_INV   0
    #define SPINDLE_PWM_AF_REMAP    0
  #endif
#endif


This seemed to work however m3 s0 turns on laser, m3 s1 or greater
turns it off. And yours above did not work(was time_n =1, a typo?).
Getting closer. Thx

**Reply by terjeio** [reply]
> m3 s0 turns on laser, m3 s1 or greater turns it off.

You have to invert the enable signal? `$16=1`.

> (was time_n =1, a typo?).

Where is that code?

**Comment by customsolutions** [comment]
I got it figured out. Like you said $16 to invert. In driver.h added the
elsif for pb9 and used 4,4,0,0 for tim,ch,inv,etc. Also now that i have the
board tested with $35 set also, I can return to machine for further
testing. Hope to move forward with mechanic mods and then stm32f466. We can
chat more when you get back if you want to add changes to main.
Thanks for your help.
